src/webgpu/shader/validation/expression/call/builtin/dot.spec.ts

```diff
@@ -0,0 +1,189 @@
+import {
+  Type,
+  kConvertableToFloatVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<any> {
+  switch (type) {
+    case Type.f32:
+      return quantizeToF32;
+    case Type.f16:
+      return quantizeToF16;
+    case Type.abstractInt:
+      return BigInt;
+    default:
+      return Number;
+  }
+}
+
+export interface CheckFunc {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (num: any): boolean;
+}
+
+function checkFunctionForScalarType(type: ScalarType): CheckFunc {
+  switch (type) {
+    case Type.abstractInt:
+      return (num: bigint) => !kValue.i64.isOOB(num);
+    default:
+      return (num: number) => Number.isFinite(num);
+  }
+}
+
+interface StepArgs {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [name: string]: any;
+}
+
+export interface EquationStep {
+  (args: StepArgs): StepArgs;
+}
+
+function checkEquation(type: ScalarType, args: StepArgs, steps: EquationStep[]) {
+  const quantizeFn = quantizeFunctionForScalarType(type);
+  const checkFn = checkFunctionForScalarType(type);
+
+  const quantizedArgs: StepArgs = {};
+  for (const key in args) {
+    quantizedArgs[key] = quantizeFn(args[key]);
+  }
+
+  for (const step of steps) {
+    const stepArgs = step(quantizedArgs);
+
+    for (const key in stepArgs) {
+      quantizedArgs[key] = quantizeFn(stepArgs[key]);
+      if (!checkFn(quantizedArgs[key])) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() never errors
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const scalarType = scalarTypeOf(kValidArgumentTypes[t.params.type]);
+
+    // Dot product equation: (a[0]*b[0]) + (a[1]*b[1]) + ... (a[N]*b[N])
```

Comment:
Because we use a single scalar replicated for all elements of the vector (splat), order of the addition doesn't need to be considered. Cool.